datalake/src/main/java/com/sequenceiq/datalake/service/recovery/RecoveryService.java

```diff
@@ -0,0 +1,53 @@
+package com.sequenceiq.datalake.service.recovery;
+
+import com.sequenceiq.cloudbreak.api.endpoint.v4.dto.NameOrCrn;
+import com.sequenceiq.sdx.api.model.SdxRecoverableResponse;
+import com.sequenceiq.sdx.api.model.SdxRecoveryRequest;
+import com.sequenceiq.sdx.api.model.SdxRecoveryResponse;
+
+/**
+ * Defines a service that can be used to recover CB deployed infrastructure.
+ *
+ * Services that can recovery failed infrastructure must be able to <em>validate</em> that a recovery action is appropriate and
+ * be able to <em>trigger</em> the recovery.
+ *
+ * The recovery request should be a value object representing an HTTP response object.
+ *
+ */
+public interface RecoveryService {
+    /**
+     * Starts recovery of the CB infrastructure related to the recovery request.
+     *
+     * This should validate that recovery is appropriate before triggerng the recovery.
+     *
+     * Trigger recovery may start a Flow or FlowChain to perform the recovery operations.
+     *
+     * Validation should have already been preformed before calling
+     *
+     * @param userCrn the CRN of the user requesting recovery
+     * @param nameOrCrn the name or CRN of the resource to recover
+     * @param upgradeRecoveryRequest detailed information about the recovery
+     * @return a response containing the identifier of the triggered recovery Flow
+     */
+    SdxRecoveryResponse triggerRecovery(String userCrn, NameOrCrn nameOrCrn, SdxRecoveryRequest recoveryRequest);
```

Comment:
why do we need nameOrCrn? I would prefer to use only crn fro internal logic, since at some point we need to find out crn anyway (for example during authorization or during other logic)

Reply:
and why do we need userCrn? It is present always on ThreadLocal, you can use it everwhere: `ThreadBasedUserCrnProvider.getUserCrn()`

Reply:
these questions applies for every method in this class